### PR TITLE
Spherical source

### PR DESCRIPTION
**Pull request by eepeterson**
This PR partially addresses issue #1394 and is just a simple addition of a spatial source distribution specified by independent distributions in r,theta,phi spherical coordinates, using the standard conventions described here. https://en.wikipedia.org/wiki/Spherical_coordinate_system#/media/File:3D_Spherical.svg

I haven't written any tests for this code and I'm not sure they are necessary, but any pointers with regards to that would be greatly appreciated. I grepped through the code looking for all the necessary files to change, but it's definitely possible I missed some. I also tried to follow all the development guidelines in the docs so please let me know if there is a part of the process that I'm missing or forgetting.

Disclaimer: I've never written code in C++ (and still haven't since this PR is mostly copy and paste), but hopefully I'll be able to get up to speed quickly.

### CODE REVIEW

**Comment by pshriwise** [comment]
I think some of the changes related to `output_summary` are coming from PR #1390 - we'll probably want to remove those to get a clean look at the intended changes.

**Comment by pshriwise** [comment]
Thanks for this submission @eepeterson! 

I think some of the changes related to `output_summary` are coming from PR #1390 - we'll probably want to remove those to get a clean look at the intended changes.

At first glance, it looks pretty good!  I'll take a closer look tomorrow.

**Comment by eepeterson** [comment]
@pshriwise Let me know if that is better.

**Comment by pshriwise** [comment]
Yup!

File: include/openmc/distribution_spatial.h [hunk 1]
```diff
@@ -37,6 +37,24 @@ class CartesianIndependent : public SpatialDistribution {
   UPtrDist z_; //!< Distribution of z coordinates
 };
 
+//==============================================================================
+//! Distribution of points specified by independent distributions in
+//r,theta,phi
```
Comment (smharper) [inline include/openmc/distribution_spatial.h]: We use Doxygen to style our comments like these. In this case, you can fit the entire sentence on one line (and still keep it under 80 charactrers) which will make Doxygen happy.

But for future reference: every line of the comment needs to start with `//! `. Also it's usually best to try making a one-line description of a class so Doxygen will interpret it as a "brief" description. (A longer description can be added after a blank line. See neighbor_list.h for an example.)

File: openmc/stats/multivariate.py [hunk 2]
```diff
@@ -375,6 +377,103 @@ def from_xml_element(cls, elem):
...
+    phi- components are sampled independently from one another.
+
+    Parameters
+    ----------
+    r : openmc.stats.Univariate
+        Distribution of r-coordinates
+    theta : openmc.stats.Univariate
+        Distribution of theta-coordinates
+    phi : openmc.stats.Univariate
+        Distribution of phi-coordinates
```
Comment (smharper) [inline openmc/stats/multivariate.py]: I think we should add some helpful reminders here of the sort:
```
    theta : openmc.stats.Univariate
        Distribution of theta-coordinates (angle relative to the z-axis)
    phi : openmc.stats.Univariate
        Distribution of phi-coordinates (azimuthal angle)
```

**Comment by smharper** [comment]
Oh, and also there are two documentation-related changes as well. Take a look at "docs/source/io_formats/settings.rst" around line 431. (This file documents our settings.xml syntax.) Add a sentence there for spherical.

We have settings.rnc and settings.rng files used for XML validation. Make the following change to the rnc file:
```patch
--- a/src/relaxng/settings.rnc
+++ b/src/relaxng/settings.rnc
@@ -75,7 +75,10 @@ element settings {
             attribute parameters { list { xsd:double+ } })? &
           element x { distribution }? &
           element y { distribution }? &
-          element z { distribution }?
+          element z { distribution }? &
+          element r { distribution }? &
+          element theta { distribution }? &
+          element phi { distribution }?
         }? &
         element angle {
           (element type { xsd:string } | attribute type { xsd:string }) &
```
And then generate a new rng file by calling `trang settings.rnc settings.rng`

**Comment by eepeterson** [comment]
Thanks for the helpful feedback @smharper! I believe I have incorporated all the changes you suggested, but let me know if something is missing still from your point of view.

File: include/openmc/distribution_spatial.h [hunk 3]
```diff
@@ -37,6 +37,23 @@ class CartesianIndependent : public SpatialDistribution {
...
+public:
+  explicit SphericalIndependent(pugi::xml_node node);
+
+  //! Sample a position from the distribution
+  //! \return Sampled position
+  Position sample() const;
+private:
+  UPtrDist r_; //!< Distribution of x coordinates
+  UPtrDist theta_; //!< Distribution of y coordinates
+  UPtrDist phi_; //!< Distribution of z coordinates
```
Comment (paulromano) [inline include/openmc/distribution_spatial.h]: Trailing comments need to be changed

File: openmc/stats/multivariate.py [hunk 4]
```diff
@@ -375,6 +377,103 @@ def from_xml_element(cls, elem):
         return cls(x, y, z)
 
 
+class SphericalIndependent(Spatial):
+    """Spatial distribution with independent r, theta, and phi distributions.
+
+    This distribution allows one to specify a coordinates whose r-, theta-, and
+    phi- components are sampled independently from one another.
```
Comment (paulromano) [inline openmc/stats/multivariate.py]: You can use ```:math:`r` ```, ```:math:`\theta` ```, and ```:math:`\phi` ``` to render the correct mathematical symbols in the documentation.

File: openmc/stats/multivariate.py [hunk 5]
```diff
@@ -375,6 +377,103 @@ def from_xml_element(cls, elem):
...
+    r : openmc.stats.Univariate
+        Distribution of r-coordinates
+    theta : openmc.stats.Univariate
+        Distribution of theta-coordinates (angle relative to the z-axis)
+    phi : openmc.stats.Univariate
+        Distribution of phi-coordinates (azimuthal angle)
+
+    """
+
+
```
Comment (paulromano) [inline openmc/stats/multivariate.py]: Remove extra blank line here

File: src/distribution_spatial.cpp [hunk 6]
```diff
@@ -51,6 +51,54 @@ Position CartesianIndependent::sample() const
...
+{
+  // Read distribution for r coordinate
+  if (check_for_node(node, "r")) {
+    pugi::xml_node node_dist = node.child("r");
+    r_ = distribution_from_xml(node_dist);
+  } else {
+    // If no distribution was specified, default to a single point at r=0
+    double x[] {0.0};
+    double p[] {1.0};
+    r_ = UPtrDist{new Discrete{x, p, 1}};
```
Comment (paulromano) [inline src/distribution_spatial.cpp]: Let's use `make_unique` instead:
```suggestion
    r_ = std::make_unique<Discrete>(x, p, 1);
```

File: src/distribution_spatial.cpp [hunk 7]
```diff
@@ -51,6 +51,54 @@ Position CartesianIndependent::sample() const
...
+
+  // Read distribution for theta coordinate
+  if (check_for_node(node, "theta")) {
+    pugi::xml_node node_dist = node.child("theta");
+    theta_ = distribution_from_xml(node_dist);
+  } else {
+    // If no distribution was specified, default to a single point at theta=0
+    double x[] {0.0};
+    double p[] {1.0};
+    theta_ = UPtrDist{new Discrete{x, p, 1}};
```
Comment (paulromano) [inline src/distribution_spatial.cpp]:
```suggestion
    theta_ = std::make_unique<Discrete>(x, p, 1);
```

File: src/distribution_spatial.cpp [hunk 8]
```diff
@@ -51,6 +51,54 @@ Position CartesianIndependent::sample() const
...
+
+  // Read distribution for phi coordinate
+  if (check_for_node(node, "phi")) {
+    pugi::xml_node node_dist = node.child("phi");
+    phi_ = distribution_from_xml(node_dist);
+  } else {
+    // If no distribution was specified, default to a single point at phi=0
+    double x[] {0.0};
+    double p[] {1.0};
+    phi_ = UPtrDist{new Discrete{x, p, 1}};
```
Comment (paulromano) [inline src/distribution_spatial.cpp]:
```suggestion
    phi_ = std::make_unique<Discrete>(x, p, 1);
```

File: tests/regression_tests/source/test.py [hunk 9]
```diff
@@ -29,9 +29,13 @@ def _build_inputs(self):
         x_dist = openmc.stats.Uniform(-3., 3.)
         y_dist = openmc.stats.Discrete([-4., -1., 3.], [0.2, 0.3, 0.5])
         z_dist = openmc.stats.Tabular([-2., 0., 2.], [0.2, 0.3, 0.2])
+        r_dist = openmc.stats.Uniform(2., 3.)
+        theta_dist = openmc.stats.Discrete([pi/4, pi/2, 3*pi/4], [0.3, 0.4, 0.3])
+        phi_dist = openmc.stats.Uniform(0.0,2*pi)
```
Comment (paulromano) [inline tests/regression_tests/source/test.py]:
```suggestion
        phi_dist = openmc.stats.Uniform(0.0, 2*pi)
```

**Comment by eepeterson** [comment]
> Nice work, @eepeterson! One thing that comes to mind is that it would be nice if the user could specify the center of the sphere rather than being stuck with the default (0, 0, 0).

Thanks, and that's a good point. I will add that. General question though: Should I bother squashing commits or will you guys squash them on merge?

**Comment by paulromano** [comment]
@eepeterson We do not squash upon merge, but feel free to squash/rebase/do whatever you want on your own branch and force push.

**Comment by eepeterson** [comment]
@smharper - sounds good I changed out the x0, y0, z0 kwargs for the origin kwarg.

File: src/distribution_spatial.cpp [hunk 10]
```diff
@@ -51,6 +51,73 @@ Position CartesianIndependent::sample() const
...
+  }
+
+  // Read sphere center coordinates
+  if (check_for_node(node, "origin")) {
+    auto origin = get_node_array<double>(node, "origin");
+    if (origin.size() == 3) {
+      origin_ = origin;
+    } else {
+      std::stringstream err_msg;
+      err_msg << "Origin for sphere must be length 3";
```
Comment (smharper) [inline src/distribution_spatial.cpp]:
```suggestion
      err_msg << "Origin for spherical source distribution must be length 3";
```

File: src/distribution_spatial.cpp [hunk 11]
```diff
@@ -51,6 +51,73 @@ Position CartesianIndependent::sample() const
...
+  }
+
+}
+
+Position SphericalIndependent::sample() const
+{
+  double r = r_->sample();
+  double theta = theta_->sample();
+  double phi = phi_->sample();
+  double x = r*sin(theta)*cos(phi) + origin_[0];
```
Comment (paulromano) [inline src/distribution_spatial.cpp]: Slightly prettier:
```suggestion
  double x = r*sin(theta)*cos(phi) + origin_.x;
```

File: src/distribution_spatial.cpp [hunk 12]
```diff
@@ -51,6 +51,73 @@ Position CartesianIndependent::sample() const
...
+  double y =  r*sin(theta)*sin(phi) + origin_[1];
```
Comment (paulromano) [inline src/distribution_spatial.cpp]:
```suggestion
  double y =  r*sin(theta)*sin(phi) + origin_.y;
```

File: src/distribution_spatial.cpp [hunk 13]
```diff
@@ -51,6 +51,73 @@ Position CartesianIndependent::sample() const
...
+  double z = r*cos(theta) + origin_[2];
```
Comment (paulromano) [inline src/distribution_spatial.cpp]:
```suggestion
  double z = r*cos(theta) + origin_.z;
```

**Comment by paulromano** [comment]
Congrats on your first PR, @eepeterson!

**Pull request merged**